File: core/src/main/java/com/natchuz/hub/core/context/ServerContext.java
package com.natchuz.hub.core.context;

import java.util.List;

import com.natchuz.hub.protocol.state.StateDatabase;
import com.natchuz.hub.core.map.MapRepository;
import com.natchuz.hub.core.modules.Module;
import com.natchuz.hub.core.profile.ProfileRepo;
import com.natchuz.hub.core.profile.UserProfile;
import com.natchuz.hub.core.proxy.ProxyBackend;

/**
 * ServerContext is an abstract factory for a different backends, starting from simple Bungeecord client to map loading etc.
 * This will be used in future to create testing context that will fake all backends, to make gamemode testing easier.
 */
public interface ServerContext {

    /**
     * Creates {@link MapRepository map repository} for this context
     */
    MapRepository createMapRepository();

    /**
     * Creates a {@link ProxyBackend proxy backaend}
     */
    ProxyBackend createProxyBackend();

    /**
     * Create a {@link StateDatabase state database}
     */
    StateDatabase createStateDatabase();

    /**
     * Create a {@link ProfileRepo profile repo} to load users from
     *
     * @param <T> type of repo
     */
    <T extends UserProfile> ProfileRepo<T> createProfileRepo(Class<T> tClass);

    /**
     * Creates a list of {@link Module modules} required to be registered in this network context
     */
    List<Module> createModules();

}

File: core/src/main/kotlin/com/natchuz/hub/core/context/ServerContext.kt
package com.natchuz.hub.core.context

import com.natchuz.hub.core.map.MapRepository
import com.natchuz.hub.core.modules.Module
import com.natchuz.hub.core.proxy.ProxyBackend
import com.natchuz.hub.core.service.PlayerService
import com.natchuz.hub.core.user.UserService

/**
 * ServerContext is an abstract factory for a different backends, starting from simple Bungeecord client to map loading etc.
 * This will be used in future to create testing context that will fake all backends, to make gamemode testing easier.
 */
interface ServerContext {
    /**
     * Creates [map repository][MapRepository] for this context
     */
    fun createMapRepository(): MapRepository

    /**
     * Creates a [proxy backaend][ProxyBackend]
     */
    fun createProxyBackend(): ProxyBackend

    /**
     * To be documented
     */
    fun createPlayerService(): PlayerService

    /**
     * Create user service
     */
    fun createUserService(): UserService

    /**
     * Creates a list of [modules][Module] required to be registered in this network context
     */
    fun createModules(): List<Module>
}